### PR TITLE
fix rawOverlap calc, add pctOverlapOfInput calc

### LINKED ISSUE #2712: KNNClassifier distance calculations
The [`rawOverlap` distance metric](https://github.com/numenta/nupic/blob/master/src/nupic/algorithms/KNNClassifier.py#L791-797) in the KNNClassifier does not calculate distance as specified [in the docstring](https://github.com/numenta/nupic/blob/master/src/nupic/algorithms/KNNClassifier.py#L102-104). Rather, it calculates a percent overlap of the input.

### PR DESCRIPTION
**Pull request by BoltzmannBrain**
Fixes #2712 by correcting the `rawOverlap` metric, and adding a `pctOverlapOfInput` metric.

### CODE REVIEW

**Comment by subutai** [comment]
If you are correcting the overlap calculation, please include at least one test case that highlights the use case that is now fixed.

**Comment by BoltzmannBrain** [comment]
@subutai wasn't ready for review yet :), but now is.

**Comment by BoltzmannBrain** [comment]
Also, the `rawOverlap` still isn't necessarily a raw overlap. As the docstring specifies, it's `(input width) - (# overlapping bits)`. Wouldn't a raw overlap just be `(# of overlapping bits)`?

**Comment by subutai** [comment]
> Also, the rawOverlap still isn't necessarily a raw overlap. As the docstring specifies, it's (input width) - (# overlapping bits). Wouldn't a raw overlap just be (# of overlapping bits)?

Yes, it should be, but with KNN it picks the closest vectors (i.e. smallest distance score). Using (input width) - (# overlapping bits) makes it so that the lower scores have higher overlap.

File: src/nupic/algorithms/KNNClassifier.py [hunk 1]
```diff
@@ -95,19 +95,22 @@ class labels
...
                 When distanceNorm is 1, this is the manhattan distance
                 In general: sum(abs(x-proto) ^ distanceNorm) ^ (1/distanceNorm)
         "rawOverlap": Only appropriate when inputs are binary. This computes:
                 (width of the input) - (# bits of overlap between input
                 and prototype).
-        "pctOverlapOfLarger": Only appropriate for binary inputs. This computes
+        "pctOverlapOfInput": Only appropriate for binary inputs. This computes
                 1.0 - (# bits overlap between input and prototype) /
-                        max(# bits in input, # bits in prototype)
+                        (# ON bits in input)
```
Comment (subutai) [inline src/nupic/algorithms/KNNClassifier.py]: Don't you mean ((width of input) - (# bits of overlap)) / (width of input)?

File: tests/unit/nupic/algorithms/knn_classifier_test.py [hunk 2]
```diff
@@ -30,6 +30,67 @@
...
+    protoB = np.array([20, 28, 30], dtype=np.int32)
+    
+    classifier.learn(protoA, 0, isSparse=dimensionality)
+    classifier.learn(protoB, 0, isSparse=dimensionality)
+    
+    input = np.zeros(dimensionality)
+    input[:4] = 1.0
+    
+    # Test l2 norm metric
+    _, _, dist, _ = classifier.infer(input)
```
Comment (subutai) [inline tests/unit/nupic/algorithms/knn_classifier_test.py]: Nice test cases. Please also include edge cases, i.e. when distance should be 0 or the largest possible.

**Comment by subutai** [comment]
Left some feedback. Otherwise :+1:

File: src/nupic/algorithms/KNNClassifier.py (hunk 1 above)
Comment (BoltzmannBrain) [inline src/nupic/algorithms/KNNClassifier.py]: No because you're first dividing by the width of input, giving a value between 0.0 (no overlapping bits) and 1.0 (all bits of the input overlap). We want to flip this, where the minimum distance is maximal overlap, so we subtract from 1.0.

**Comment by BoltzmannBrain** [comment]
@subutai I addressed your comments. Please merge if it's good to go.

**Comment by BoltzmannBrain** [comment]
@scottpurdy Subutai may have left for his vacation, so please take a look.

**Comment by BoltzmannBrain** [comment]
or @rhyolight

File: tests/unit/nupic/algorithms/knn_classifier_test.py [hunk 3]
```diff
@@ -30,6 +30,95 @@
...
+    # input0 is used to test that the distance from a point to itself is 0
+    input0 = np.zeros(dimensionality)
+    input0[protoA] = 1.0
+
+    # Test l2 norm metric
+    _, _, dist, _ = classifier.infer(input)
+    l2Distances = [0.65465367,  1.0]
+    for actual, predicted in zip(l2Distances, dist):
+      self.assertAlmostEqual(actual, predicted, places=5,
+        msg="l2 distance norm is not calculated as expected.")
```
Comment (scottpurdy) [inline tests/unit/nupic/algorithms/knn_classifier_test.py]: If you want to break like this then start the first argument on a new line. See [PEP-8](https://www.python.org/dev/peps/pep-0008/#indentation).

**Comment by scottpurdy** [comment]
:+1:

**Comment by BoltzmannBrain** [comment]
Thanks @scottpurdy, please merge if ready.
